src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/ConditionalWeakTable.cs

```diff
@@ -485,6 +480,56 @@ public KeyValuePair<TKey, TValue> Current
                     return _current;
                 }
             }
+        }
+
+        /// <summary>Provides an enumerator for the table.</summary>
+        private sealed class Enumerator : IEnumerator<KeyValuePair<TKey, TValue>>
+        {
+            private EnumeratorCore _enumeratorCore;
+
+            public Enumerator(ConditionalWeakTable<TKey, TValue> table) => _enumeratorCore = new(table);
+
+            ~Enumerator()
+            {
+                Dispose();
+            }
+
+            public void Dispose() => _enumeratorCore.Dispose();
+
+            public bool MoveNext() => _enumeratorCore.MoveNext();
+
+            public KeyValuePair<TKey, TValue> Current => _enumeratorCore.Current;
+
+            object? IEnumerator.Current => Current;
+
+            public void Reset() { }
+        }
+
+        /// <summary>Gets an enumerable used to enumerate the values on the stack without allocation.</summary>
+        // This API is unsafe to use if the user neglects to call Dispose, which is why
+        // we can only use this internally where we can enforce proper foreach usage.
+        // It's also unsafe to use in async functions, even with proper foreach, because there's
+        // a chance that the async will never continue, in which case we need to rely on the finalizer in the regular enumerator.
+        internal RefEnumerable EnumerateOnStack() => new RefEnumerable(this);
+
+        /// <summary>Provides an enumerable used to enumerate the values on the stack without allocation.</summary>
+        internal readonly ref struct RefEnumerable(ConditionalWeakTable<TKey, TValue> table)
+        {
+            public RefEnumerator GetEnumerator() => new(table);
+        }
+
+        /// <summary>Provides an enumerator for the table that can be used on the stack without allocation.</summary>
+        internal ref struct RefEnumerator : IEnumerator<KeyValuePair<TKey, TValue>>
```

Comment:
See https://github.com/dotnet/runtime/issues/53296 for a long discussion about why it is not the best idea to expose struct enumerator for ConditionalWeakTable

Reply:
Ok, I see you have mentioned in a comment - but that does not make it a good idea.

Reply:
I'm well aware of the dangers, which is why it's internal only. I left a lengthy comment about it on the `EnumerateOnStack()` method.

Reply:
We are not interested in spreading more unsafe or otherwise dangerous code throughout the libraries. We believe we have too much of it and we are actively working on reducing it.